### ISSUE TITLE
upgrade from 16.0.3 fails silently

### ISSUE THREAD

**Issue by bestouff** [nextcloud/server#17385]
When trying to upgrade from 16.0.3, the updater fails by saying Nextcloud is already at the latest version:
```
$ php updater.phar --no-interaction
Nextcloud Updater - version: v16.0.0beta3-4-gc7a4074

Current version is 16.0.3.

Update to Nextcloud 16.0.4 available. (channel: "production")
Following file will be downloaded automatically: https://download.nextcloud.com/server/releases/nextcloud-16.0.4.zip
Open changelog ↗

Updater run in non-interactive mode.

Continue update

Info: Pressing Ctrl-C will finish the currently running step and then stops the updater.

[✔] Check for expected files
[✔] Check for write permissions
[✔] Create backup
[✔] Downloading
[✔] Verify integrity
[✔] Extracting
[✔] Enable maintenance mode
[✔] Replace entry points
[✔] Delete old files
[✔] Move new files in place
[✔] Done

Update of code successful.
Updater run in non-interactive mode - will start "occ upgrade" now.

The current PHP memory limit is below the recommended value of 512MB.
Nextcloud is already latest version

Updater run in non-interactive mode - will disable maintenance mode now.
The current PHP memory limit is below the recommended value of 512MB.
Maintenance mode already disabled

Maintenance mode is disabled
```

**Comment by j-ed** [comment]
> The current PHP memory limit is below the recommended value of 512MB.

I would assume that the update will succeed, once you've correctly adjusted the [PHP memory limit](https://help.nextcloud.com/t/message-php-memory-limit-below-recommended-value-of-512-mb-wont-vanish-die-meldung-php-speichergrenze-liegt-unterhalb-des-empfohlenen-wertes-von-512mb-verschwindet-nicht/58374) parameter.

BTW, this kind of support requests should usually be addressed in the [**Nextcloud help forum**](https://help.nextcloud.com/) and not in this bug tracker 😉

**Comment by bestouff** [comment]
This parameter is fixed to 128M by my provider and is not modifiable. Does that mean Nextcloud will be stuck to 16.0.3 ?  
Curiously the other upgrades worked without problems.

**Comment by kesselb** [comment]
Whats the output of `php -v`  and does the version match with the one used by the webserver?

**Comment by bestouff** [comment]
```
PHP 7.3.6 (cli) (built: Jun 26 2019 15:15:06) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.6, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.6, Copyright (c) 1999-2018, by Zend Technologies
```
Yes there's only one PHP instll on these servers (Gandi Simple Hosting)

**Comment by bestouff** [comment]
> I would assume that the update will succeed, once you've correctly adjusted the PHP memory limit parameter.  

Why would a *recommended* value prevent an upgrade from happening ?

**Comment by j-ed** [comment]
> Why would a recommended value prevent an upgrade from happening ?

Eventually because PHP functions are called during the update which could require more memory to proceed?! Why should the memory requirement be checked if it isn't required?

**Comment by bestouff** [comment]
Ok, so why isn't there an error message to this effect ?

**Comment by kesselb** [comment]
Have you tried it without `--no-interaction`?

**Comment by bestouff** [comment]
Yes, both with and without (the log in this issue is with `--no-interaction`).

**Comment by kesselb** [comment]
https://github.com/nextcloud/server/issues/16888
https://github.com/nextcloud/updater/issues/243

Hmm. I think the updater starts with an old state somehow. 

In my data folder there is a updater-$somerandomcharacters folder. Is there a file .step-previous-update and if so could you share the content? It looks like this on my instance: `{"state":"end","step":12}`

**Comment by bestouff** [comment]
```sh
$ cat .step-previous-update ; echo
{"state":"end","step":10}
$ cat .step ; echo
{"state":"end","step":12}
```

**Comment by bestouff** [comment]
```sh
$ cat version.php 
<?php 
$OC_Version = array(16,0,3,0);
$OC_VersionString = '16.0.3';
$OC_Edition = '';
$OC_Channel = 'stable';
$OC_VersionCanBeUpgradedFrom = array (
  'nextcloud' => 
  array (
    '15.0' => true,
    '16.0' => true,
  ),
  'owncloud' => 
  array (
  ),
);
$OC_Build = '2019-07-09T12:01:37+00:00 185087e9bc46c1ef08e165e1ac979827dc05d469';
$vendor = 'nextcloud';
```

**Comment by kesselb** [comment]
OK. For some reason the files are not replaced. I think we have another report somewhere but don't know where. A workaround is to do a manual upgrade: https://docs.nextcloud.com/server/16/admin_manual/maintenance/manual_upgrade.html

Please ensure that you have a working backup. I'm not sure how accurate the manual upgrade guide is. For example some of the chmod commands look questionable. I guess they work on a debian linux installation but the right permissions for the file can vary.

**Comment by bestouff** [comment]
Ok, I see. What if I just remove the `.step*` files, or even the `upgrade*` directory ?

**Comment by kesselb** [comment]
The `updater-*` directory contains some backups. `.step` file should be delete at the end of the update anyway (https://github.com/nextcloud/updater/blob/96234eb99dd5b887350e63d1b0815a978395bf65/lib/Updater.php#L966). I'm not sure about `.step-previous-update` sorry :see_no_evil:
